### PR TITLE
Add data type error detection notebook

### PR DESCRIPTION
**Pull request by SankBad**
This PR will add a data type error detection notebook for public repo of configuration file analysis.

This PR is a continuation of PR #2  and I made suggested changes.

### CODE REVIEW

**Comment by SankBad** [comment]
Hi @MichaelClifford  , @Shreyanand  , 

Thanks for the feedback on PR #2 

1) There is a bug in the insight configuration parser. If we have a file with inline comments, the parser is not working as expected. I have created an issue for the same. (https://github.com/RedHatInsights/insights-core/issues/2777)
2) Certain conditions for assigning data types are strict. Such as if data type is size i.e. 14MB, 1G etc. it didnt accept the integer values. If most of the values are 0||1 consider the data type as 0||1; It will create more warnings(FP), but it might catch the error. 
**I ease all these kinds of conditions in this commit.**  
3) Finally, I found it's difficult to include all the conditions for determining the data type from given values using RegEx. Like filename [conventions](https://en.wikipedia.org/wiki/Filename), a file-path convention for different OS(in this data many configuration files are configured for windows). There are always other corner cases for determining data types such as urls, IP address, ipnetwork etc.

Hence, I removed all the regex base data inference techniques except for size determination as it is a simple regex query. I replaced all the techniques by using python libraries such as ipaddress, pathvalidatior, urllib etc. 
I made a certain modification in those functions according to our needs. 


> Sanket, did we labels for this data?

No, this data is not labeled i.e we don't know it is type error or not. Basically, all the files are collected from the web, including Q&A forums and the official mailing lists of the software such as StackOverflow. Many time issues arise because users manually anonymized the path of files, port numbers etc. Most of the urls are not anymore valid and we just know the one word [issue](https://github.com/tianyin/configuration_datasets/blob/master/configfiles/mysql/v-5.x.metadata.csv) 

After making all these changes I mentioned above, I validated all the rules and detected errors by myself. All of them are making sense. Percentage of errors are currently 10.63 most and most of the errors were being fixed because of point number 2: many time users provided size in integer format. 10-15 files. an earlier total error was 51.

Currently, most of the detected issues are caused by the bug in the insight parser that I mentioned in point number 2. Hence, I am not sure is it fine to show all the errors in the notebook.

**Comment by Shreyanand** [comment]
> Certain conditions for assigning data types are strict. Such as if data type is size i.e. 14MB, 1G etc. it didnt accept the integer values. If most of the values are 0||1 consider the data type as 0||1; It will create more warnings(FP), but it might catch the error.

Could you please elaborate on this?

> Currently, most of the detected issues are caused by the bug in the insight parser that I mentioned in point number 2. Hence, I am not sure is it fine to show all the errors in the notebook.

The bug is mentioned in point 1, I think I'm missing something here...

**Comment by SankBad** [comment]
> > Certain conditions for assigning data types are strict. Such as if data type is size i.e. 14MB, 1G etc. it didnt accept the integer values. If most of the values are 0||1 consider the data type as 0||1; It will create more warnings(FP), but it might catch the error.
> 
> Could you please elaborate on this?

For certain keys such as innodb_buffer_pool_size datatype is size (128M, 2G, 1K etc) but this particular entry also accepts integer value. i.e. users can give input as 134217728 as well as 128 M. both of which are valid. The previous implementation throws errors if someone passes integer values. 

For a certain key, if 95% values are 0||1 we will consider the datatype as 0||1. but if someone passes the boolean value ( ['on', 'off', 'true', 'false']) it will trigger the error. 

> > Currently, most of the detected issues are caused by the bug in the insight parser that I mentioned in point number 2. Hence, I am not sure is it fine to show all the errors in the notebook.
> 
> The bug is mentioned in point 1, I think I'm missing something here...

Sorry, that was a typo. I mean point number 1 :facepalm:

File: notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb [hunk 1]
```diff
@@ -0,0 +1,821 @@
+{
```
Comment (SankBad) [inline notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: <p>In the above code, data_dir is the folder where we are using all the mysql configuration files. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/configuration-files-analysis/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (SankBad) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: @MichaelClifford , In the above line of the code data_dir is the folder of all MySQL based configuration files.

Comment (MichaelClifford) [inline notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: <p>Can you move all of the imports to the same cell above?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/configuration-files-analysis/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (SankBad) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: Hi @MichaelClifford , 
I moved all of the imports to the top of the same cell.

**Comment by SankBad** [comment]
Updated folder path of mysql files according to changes in the data_downloader.py script
@MichaelClifford @Shreyanand , 
I made all the changes you can review this PR.

File: notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
```
Comment (MichaelClifford) [inline notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: <p>Place all import statements into this cell. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/configuration-files-analysis/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (SankBad) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: Sorry, I misunderstood in the starting :facepalm: . I moved all the import into the same cell.

**Review by MichaelClifford — APPROVED** [review]
LGTM :+1:

**Review by Shreyanand — CHANGES_REQUESTED** [review]
@SankBad The notebook is in a much better shape now :+1: 

I have added a comment for a minor code change.

I was wondering about the result. The "MySQL" string instead of filepath error is a good example. You mentioned most of the other errors are due to parsers and not due to a human error? What percentage is parser induced (out of the 10% that you see?)
Would it make sense to separately report them, or in fact completely remove the parser induced errors from the results...

File: notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb (hunk 1 above)
Comment (Shreyanand) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: <p>&nbsp;&nbsp;if not scheme or scheme:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;return False</p><p> </p><p>Is this code correct?</p>

Comment (SankBad) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: my bad. I made some mistake in final version.
it should be `if not scheme or not domain:`

Comment (SankBad) [reply notebooks/Misconfiguration_detection_framework_for_data_type_errors.ipynb:1]: I fixed this issue

**Comment by SankBad** [comment]
@Shreyanand , 

I removed a dependency from the insights parser and using configparser package in python. 
by default, it also gives an error for inline comments but we have a parameter `inline_comment_prefixes` which fixes these issues. All the issues related to inline comments are fixed now. 

You can review this PR.

**Comment by Shreyanand** [comment]
> @Shreyanand ,
> 
> I removed a dependency from the insights parser and using configparser package in python.
> by default, it also gives an error for inline comments but we have a parameter `inline_comment_prefixes` which fixes these issues. All the issues related to inline comments are fixed now.
> 
> You can review this PR.

So the 6.98% now are all human errors? I'm curious, what are they like? Are they similar to "MySQL" error?

**Comment by SankBad** [comment]
@Shreyanand 
These are the detected 19 errors. Out of which 11 are indeed errors.

We detected data type of keyword nice as 0||1 or bool because the majority of files have these values but sometimes values are also integer.

We detected datatype of query_cache_type as 0||1 but sometime it also can take values as 2. Therefore, we are getting those errors. Documentation also said that this variable takes values as 0, 1, 2

Last, It seems like when datatype is size we are just allowing int data type. We should also allow 0||1. (I can make this change)




```

0700f8556bfa82b829425cd3fcc0c296.cnf
Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as '0||1'
0700f8556bfa82b829425cd3fcc0c296.cnf
Type_Error: Data type of key 'read_rnd_buffer_size' should be ['int', 'size'] but its given as '0||1'
11b9fa75593f39219532d16270c0d5d5.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
205d181d85f71fbf0cd8e97294418e7b.cnf
Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
35df17d3f58e25ef7c616871f000b657.cnf
Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
36517fcad728674ebebd7c4e8c24928d.cnf
Type_Error: Data type of key 'wait_timeout' should be ['int', '0||1', 'size'] but its given as 'string'
36655ab464d87a7833deaac8fd21dd1d.cnf
Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
459f246424a34a2a9908406104cb8334.cnf
Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
597ce35462a0ff54f42dba47e3a1b985.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
749645bdd680596da71a139cfb0e000a.cnf
Type_Error: Data type of key 'query_cache_type' should be ['0||1', 'bool'] but its given as 'int'
983d11671a378c03c43355828bf71e8e.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
993af96aa93a8fed584d9230f4098dcc.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
c1040f2e32a66812871720cb54f47db5.cnf
Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
c194e559589371465c627b3e1bd088de.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
d822d1d46ff0485364618b2f7bc9c92d.cnf
Type_Error: Data type of key 'query_cache_type' should be ['0||1', 'bool'] but its given as 'int'
da4cab5a12e152a9b5018620a9c43dc7.cnf
Type_Error: Data type of key 'query_cache_limit' should be ['int', 'size'] but its given as '0||1'
e384ee4960fb60c2d130cc79e63b57f0.cnf
Type_Error: Data type of key 'log-error' should be ['filename||filepath+filename'] but its given as 'string'
eeb8aa7c01994d2ec778fd2cba71d3bd.cnf
Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
fdb71c67a14f12a29a0fabdc3b003e44.cnf
Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as 'string'

**Comment by Shreyanand** [comment]
> @Shreyanand
> These are the detected 19 errors. Out of which 11 are indeed errors.
> 
which ones?

> We detected data type of keyword nice as 0||1 or bool because the majority of files have these values but sometimes values are also integer.
> 
> We detected datatype of query_cache_type as 0||1 but sometime it also can take values as 2. Therefore, we are getting those errors. Documentation also said that this variable takes values as 0, 1, 2
> 
> Last, It seems like when datatype is size we are just allowing int data type. We should also allow 0||1. (I can make this change)
> 
> ```
> 
> 0700f8556bfa82b829425cd3fcc0c296.cnf
> Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as '0||1'
> 0700f8556bfa82b829425cd3fcc0c296.cnf
> Type_Error: Data type of key 'read_rnd_buffer_size' should be ['int', 'size'] but its given as '0||1'
> 11b9fa75593f39219532d16270c0d5d5.cnf
> Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
> 205d181d85f71fbf0cd8e97294418e7b.cnf
> Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
> 35df17d3f58e25ef7c616871f000b657.cnf
> Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
> 36517fcad728674ebebd7c4e8c24928d.cnf
> Type_Error: Data type of key 'wait_timeout' should be ['int', '0||1', 'size'] but its given as 'string'
> 36655ab464d87a7833deaac8fd21dd1d.cnf
> Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
> 459f246424a34a2a9908406104cb8334.cnf
> Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
> 597ce35462a0ff54f42dba47e3a1b985.cnf
> Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
> 749645bdd680596da71a139cfb0e000a.cnf
> Type_Error: Data type of key 'query_cache_type' should be ['0||1', 'bool'] but its given as 'int'
> 983d11671a378c03c43355828bf71e8e.cnf
> Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
> 993af96aa93a8fed584d9230f4098dcc.cnf
> Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
> c1040f2e32a66812871720cb54f47db5.cnf
> Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
> c194e559589371465c627b3e1bd088de.cnf
> Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
> d822d1d46ff0485364618b2f7bc9c92d.cnf
> Type_Error: Data type of key 'query_cache_type' should be ['0||1', 'bool'] but its given as 'int'
> da4cab5a12e152a9b5018620a9c43dc7.cnf
> Type_Error: Data type of key 'query_cache_limit' should be ['int', 'size'] but its given as '0||1'
> e384ee4960fb60c2d130cc79e63b57f0.cnf
> Type_Error: Data type of key 'log-error' should be ['filename||filepath+filename'] but its given as 'string'
> eeb8aa7c01994d2ec778fd2cba71d3bd.cnf
> Type_Error: Data type of key 'nice' should be ['0||1', 'bool'] but its given as 'int'
> fdb71c67a14f12a29a0fabdc3b003e44.cnf
> Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as 'string'
> ```

**Comment by SankBad** [comment]
> which ones?

Please find a list
```
0700f8556bfa82b829425cd3fcc0c296.cnf
Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as '0||1'
11b9fa75593f39219532d16270c0d5d5.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
35df17d3f58e25ef7c616871f000b657.cnf
Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
36517fcad728674ebebd7c4e8c24928d.cnf
Type_Error: Data type of key 'wait_timeout' should be ['int', '0||1', 'size'] but its given as 'string'
36655ab464d87a7833deaac8fd21dd1d.cnf
Type_Error: Data type of key 'port' should be ['int', '0||1', 'size'] but its given as 'string'
597ce35462a0ff54f42dba47e3a1b985.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
983d11671a378c03c43355828bf71e8e.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
993af96aa93a8fed584d9230f4098dcc.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'filepath'
c194e559589371465c627b3e1bd088de.cnf
Type_Error: Data type of key 'socket' should be ['filename||filepath+filename'] but its given as 'string'
e384ee4960fb60c2d130cc79e63b57f0.cnf
Type_Error: Data type of key 'log-error' should be ['filename||filepath+filename'] but its given as 'string'
fdb71c67a14f12a29a0fabdc3b003e44.cnf
Type_Error: Data type of key 'read_buffer_size' should be ['int', 'size'] but its given as 'string'```

**Comment by SankBad** [comment]
@Shreyanand , 

Also, when datatype is the size we should not allow 0||1. Because all size parameters accept values as 1MB, 1KB, 1G, 1M, etc.  or 1024 etc values in bytes.

So, If we allow 0 or 1. that means 0 or 1 bytes that are not a suitable value for example:
https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html 
Where datatype is memory(sizes) minimum values are in thousands

**Comment by Shreyanand** [comment]
Cool, this list seems reasonable.
LGTM :+1:

**Pull request merged**